### PR TITLE
Issue 04

### PR DESCRIPTION
**Pull request by kolotaev**
Ruby client for Samsara SDK implementation.

### CODE REVIEW

**Comment by kolotaev** [comment]
@BrunoBonacci,

The approach, proposed by you, is much more efficient both operation-speed and memory wise: we delete with O(1) and use O(1) memory (previous one copied the whole buffer). So from this perspective  it's definitely worth to implement. The only drawback I consider it is a little bit tricky - I've spend some time struggling with edge-cases though from the very first sight its idea is quite simple.

Anyway, thanks for the the tip of a new algorithm! In the last commit you can review its implementation for Samsara ring-buffer.

**Comment by BrunoBonacci** [comment]
Great news, I'll review the changes in the next couple of days and merge the PR.
thx

**Comment by BrunoBonacci** [comment]
@kolotaev,

Excellent job! The new ring buffer is very clear and simple.
I just posted two questions in the review section on `take_snapshot` and `delete` methods.
Everything else is spot on!

File: clients/ruby/lib/samsara_sdk/ring_buffer.rb [hunk 1]
```diff
@@ -49,36 +54,71 @@ def push(value)
...
     #
-    # @param snapshot [Array<Object>] Snapshot of a buffer at some state.
-    def delete(snapshot)
+    # @param mark [Integer] high attribute that represents high-water-mark of a snapshot.
+    def delete(mark)
       @mutex.synchronize do
-        @buffer.each_index { |i| @buffer[i] = nil if @buffer[i] == snapshot[i] }
-        @count = @buffer.compact.size
-        @pointer = @size - 1 if empty?
+        @low = @high - mark > @size ? @high - @size : mark
```
Comment (kolotaev) [inline clients/ruby/lib/samsara_sdk/ring_buffer.rb]: Agree here. I'll correct this.

File: clients/ruby/lib/samsara_sdk/ring_buffer.rb [hunk 2]
```diff
@@ -49,36 +54,71 @@ def push(value)
...
+                 []
+               elsif @low == -1
+                 @buffer.compact
+               elsif @high < @size # array is not filled completely
+                 @buffer[j..i]
+               else
+                 @buffer[j..-1] + @buffer[0..i]
+               end
+
+        return data, @high
```
Comment (kolotaev) [inline clients/ruby/lib/samsara_sdk/ring_buffer.rb]: Yes. This one is much simpler for future support.
But initially I refused this approach due to the performance issue.

This is a little benchmark.
```ruby
def time
  (Time.now.to_f.round(3) * 1000).to_i
end

size = 500_000

buffer = Array.new(size) { rand(9999) }

puts 'copy by each:'
a = Array.new
start = time
size.times do |i|
  a << buffer[i]
end
puts time - start

puts 'copy by indexes:'
b = Array.new(size)
start = time
i = size / 2
b = buffer[i..size-1] + buffer[0...i]
puts time - start
```
Ruby's C manipulations with arrays using indices is faster then direct `Enum#each`.
On my machine
For 500_000 entries enum is 50 ms - C is 2 ms
For 1_000_000 entries enum is 90 ms - C is 13 ms

Ifs are for handling different cases in order to preserve FIFO.

We have to decide here: performance vs. clear code.

Comment (BrunoBonacci) [inline clients/ruby/lib/samsara_sdk/ring_buffer.rb]: got it, now I understand why you did so.
I generally prefer clear code over performances unless performances are a clear project constraint which limit the outcome.
However in this case the performance "hack" is limited to a single function,
so if we document well this one explaining the motivation we can leave it as it is.

We can add a comment like:
```
take_snapshot returns a new array of all elements between @low and @high copying each element. Rater than copying the elements via Ruby's Array facility we use the lower and
more efficient C array copy.
etc etc  
```

**Comment by kolotaev** [comment]
Done simplification.

**Pull request merged**